### PR TITLE
Mech update 3 (The forklifting update)

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
Adds in the following features
1)Updates mech hud to latest bay
2)Adds powering down / powering up of mechs
3)Adds mobile power generator for mechs(with sounds , no light)
 Welding fuel generator , generates a max of 20 KW of power , at a rate of 100 W per unit of fuel. Stores 200 units. Can be saboutaged if plasma is inserted into its canister
 Plasma fuel generator, generates a max of 35 KW of power , at a rate of 250 W per 50 plasma , each sheet is 1000 plasma units. Cannot be saboutaged
4)Adds the mech medical system (uses advanced trauma kits to seal all wounds automatically and fast , will fully heal limbs)
5)Adds mech towing module , mech towing can not be overridden by anyone(+ inbuilt safety against towing loops)
6)Adds the forklift mech chassis , wheels  propulsion system and the forklifting system
 The forklifting system can be used to lift anything that can move and is not anchored down. Attacking the system whilst its selected in the UI will cause it to start lifting , which will raise any lifted mob up a z-level and put the forklift platform under them(and also let them move and do actions freely). Retracting the forklift whilst a mob is sitting above it will instantly put them back in the forklift.
 The forklift clamps can be instantly exited by any mob by resisting
7)Allows mechs to be built without arms and heads , and adds support for mechs without arms / legs to work properly
8)Adds the mech toolkit , a mech mounted toolset with high tool qualities compared to the average
9)Code framework work (Adds support for objects that prevent Z-falling, Reworks how shields on mechs handle their updates, etc)
10)Adds more examine info for mechs. 
11) Fixes mech module ejection to stop having objects stuck to the screen
12)Added the MK2  sleeeper , with tramadol  , tricord and polystem to the default one(which also costs 1 platinum)
13)Makes the ballisic shield undeploy time significantly faster + adds updated sprites for deployed / not deployed.
14) Updates mech shield sprites.

 

## Why It's Good For The Game
More mech content
## Testing
Tested on local
Mech hud works as expected 
Power states do not break , and you can toggle on a mech if it has no cell as long as theres a power generator with a cell
Tested mobile generators with the mech + sound,  the louder they ran , the farther they were heard from
Tested the toolkit on doors and walls
Tested the auto-mender on a few human mobs
Tested Z-level falling code to make sure my changes did not break it
Tested ballistic shields / shield generator for new system support
Tested forklifting up a z-level, tested being forklifted 
Tested the new mech chassis and made sure it looks proper
Tested mech building without arms and legs

## Changelog
:cl:
add: Added the following mech modules : automender , sleeper MK2 , towing hook , forklifting system , engineering , toolkit, power generators(plasma and welding fuel)
add: Added the forklift chassis and wheels propulsion type
fix: Fixed mech UI breaking when removing a object by force,  fixed suffocating with a closed open chassis.
add: Updated the sprites of the mech ballistic shield and the shield generator when being deployed/undeployed and when shield is projected.
/:cl:

### CODE REVIEW

**Comment by Firefox13** [comment]
> изображение

Put some effort please

**Comment by Humonitarian** [comment]
Devdel stamp of approval. This is not a balance approval, but a content one.

If balance is a concern (hopefully not), I can look at it tomorrow.

**Comment by Firefox13** [comment]
Also images would be VERY welcome

**Comment by MLGTASTICa** [comment]
image

**Comment by Firefox13** [comment]
No, when I ask for images I mean proper screenshots and not a mis mash of 32x32 sprites. They're unreadable.

**Comment by MLGTASTICa** [comment]
image
image
image
image
image

**Comment by githubuser4141** [comment]
What does it look like if equipment is auto dyed to match limb color?

**Comment by Firefox13** [comment]
How does forklift chassis look with other mech parts?
Can human interact with whatever forklift is carrying? I remember discussing this with WastelandMain on discord and we went with grappling hook something, so you could fuck with the thing mech is carrying. Is this the case here? How does carrying a crate/locker look when facing north, is it hidden behind the mech sprite? Can you carry people, what happens if you carry someone in a closet?

**Comment by MLGTASTICa** [comment]
> How does forklift chassis look with other mech parts? Can human interact with whatever forklift is carrying? I remember discussing this with WastelandMain on discord and we went with grappling hook something, so you could fuck with the thing mech is carrying. Is this the case here? How does carrying a crate/locker look when facing north, is it hidden behind the mech sprite? Can you carry people, what happens if you carry someone in a closet?

image
image
image
image
image
image
Whatever the forklift is carrying , cannot be interacted unless the arms are lifted above. The stuff that is carrying can however be removed by attacking it on grab intent with empty hands(and is done so in 3 seconds) . People inside can also instantly resist out of the forklift if they are being carried. People who are being forklifted can see as normal but cannot act with anything on the outside / shoot guns or throw stuff.  People inside a closet that are being carried can just resist out of the closet and will be put outside.
As for other mech parts ,a forklift chassis cannot take any other legs other than the wheels type.

**Comment by Humonitarian** [comment]
> a forklift chassis cannot take any other legs other than the wheels type.

Would the sprites conflict if you could?
Can you put wheels on other mech types?

**Comment by githubuser4141** [comment]
Looks fine imo with combat and light legs:
image

**Comment by MLGTASTICa** [comment]
> > a forklift chassis cannot take any other legs other than the wheels type.
> 
> Would the sprites conflict if you could? Can you put wheels on other mech types?

yes you can put wheels on other mech types , however wheels have the least health.

**Comment by thevandie** [comment]
You're too late sonic, I am now forklift certified!

File: code/modules/mechs/equipment/utility.dm [hunk 1]
```diff
@@ -465,3 +465,595 @@
 
 /obj/item/extinguisher/mech/get_hardpoint_status_value()
 	return reagents.total_volume/max_water
+
+/obj/item/mech_equipment/power_generator
+	name = "debug power generator"
+	desc = "If you see this tell coders to fix code.!"
```
Comment (TheShown911) [inline code/modules/mechs/equipment/utility.dm:471]: Fix the interpunction first.

File: code/modules/mechs/interface/datum_HUD.dm [hunk 2]
```diff
@@ -9,6 +9,7 @@
 		"mech radio"				=	list("type"	=	/obj/screen/movable/exosuit/radio,				"loc" = "WEST:4,CENTER-3:58"),
 		"rename mech"				=	list("type"	=	/obj/screen/movable/exosuit/rename,				"loc" = "WEST:4,CENTER-3:69"),
 		"mech camera"				=	list("type"	=	/obj/screen/movable/exosuit/toggle/camera,		"loc" = "WEST:4,CENTER-3:80"),
+		"mech power switch"			=	list("type" =	/obj/screen/movable/exosuit/toggle/power_control, "loc" = "WEST:4,CENTER-3:-12"),
```
Comment (TheShown911) [inline code/modules/mechs/interface/datum_HUD.dm:12]: Line this up.

**Review by TheShown911 — APPROVED** [review]
Code looks good, can't say anything about balance or the writing. Such a big PR should be tested on the test server but there is no one to test it so i just say merge it.

File: code/modules/mechs/interface/datum_HUD.dm (hunk 2 above)
Comment (MLGTASTICa) [reply code/modules/mechs/interface/datum_HUD.dm:12]: it is lined. Check file in view/edit mode.

Comment (TheShown911) [reply code/modules/mechs/interface/datum_HUD.dm:12]: Poggers

**Comment by thevandie** [comment]
> image

oh THATS how people see the damn image changes, I just thought people were using an extension

**Comment by MLGTASTICa** [comment]
Shield sprites changed to the ones agreed by firefox to at https://discord.com/channels/255035529085583360/265411250341543936/1184743078809055252

**Pull request merged**